API/hermes_abi/hermes_abi.h

```diff
@@ -0,0 +1,270 @@
+      const char *);
+
+  HermesABIObjectOrError (*get_global_object)(HermesABIContext *);
+  HermesABIStringOrError (
+      *create_string_from_ascii)(HermesABIContext *, const char *, size_t);
+  HermesABIStringOrError (
+      *create_string_from_utf8)(HermesABIContext *, const uint8_t *, size_t);
+
+  HermesABIValue (*get_object_property_from_string)(
+      HermesABIContext *,
+      HermesABIObject,
+      HermesABIString);
+  HermesABIValue (*get_object_property_from_prop_name_id)(
+      HermesABIContext *,
+      HermesABIObject,
+      HermesABIPropNameID);
+  HermesABIVoidOrError (*set_object_property_from_string)(
+      HermesABIContext *,
+      HermesABIObject,
+      HermesABIString,
+      const HermesABIValue *);
+  HermesABIVoidOrError (*set_object_property_from_prop_name_id)(
+      HermesABIContext *,
+      HermesABIObject,
+      HermesABIPropNameID,
+      const HermesABIValue *);
+
+  HermesABIPropNameIDOrError (*create_prop_name_id_from_ascii)(
+      HermesABIContext *,
+      const char *,
+      size_t);
+  HermesABIPropNameIDOrError (*create_prop_name_id_from_utf8)(
+      HermesABIContext *,
+      const uint8_t *,
+      size_t);
+  HermesABIPropNameIDOrError (
+      *create_prop_name_id_from_string)(HermesABIContext *, HermesABIString);
+  HermesABIPropNameIDOrError (
+      *create_prop_name_id_from_symbol)(HermesABIContext *, HermesABISymbol);
+  HermesABIBoolOrError (*compare_prop_name_id)(
+      HermesABIContext *,
+      HermesABIPropNameID,
+      HermesABIPropNameID);
+
+  HermesABIBoolOrError (*is_array)(HermesABIContext *, HermesABIObject);
+  HermesABIBoolOrError (*is_array_buffer)(HermesABIContext *, HermesABIObject);
+  HermesABIBoolOrError (*is_function)(HermesABIContext *, HermesABIObject);
+  HermesABIBoolOrError (*is_host_object)(HermesABIContext *, HermesABIObject);
+  HermesABIBoolOrError (
+      *is_host_function)(HermesABIContext *, HermesABIFunction);
+
+  HermesABIValue (*call)(
+      HermesABIContext *,
+      HermesABIFunction,
+      HermesABIValue *,
+      HermesABIValue *,
+      size_t);
+
+  HermesABIFunctionOrError (*create_function_from_host_function)(
+      HermesABIContext *,
+      HermesABIPropNameID,
+      unsigned int,
+      HermesABIHostFunction *);
+  HermesABIHostFunction *(
+      *get_host_function)(HermesABIContext *, HermesABIFunction);
+
+  HermesABIPropNameIDOrError (
+      *clone_prop_name_id)(HermesABIContext *, HermesABIPropNameID);
+  HermesABIStringOrError (*clone_string)(HermesABIContext *, HermesABIString);
+  HermesABISymbolOrError (*clone_symbol)(HermesABIContext *, HermesABISymbol);
+  HermesABIObjectOrError (*clone_object)(HermesABIContext *, HermesABIObject);
+  HermesABIBigIntOrError (*clone_big_int)(HermesABIContext *, HermesABIBigInt);
+
+  size_t (
+      *get_max_utf8_length_from_string)(HermesABIContext *, HermesABIString);
+  HermesABISizeTOrError (
+      *get_utf8_from_string)(HermesABIContext *, HermesABIString, uint8_t *);
```

Comment:
Consider using a different pattern where the same function can return either number of required bytes or fill the target buffer. Otherwise, the `get_utf8_from_prop_name_id` looks quite dangerous because it does not provide the buffer size.
See [napi_get_value_string_utf8](https://github.com/nodejs/node/blob/3af7cfe7d4b55ed88a68d2a04212d72540a4951a/src/js_native_api.h#L171)

Reply:
This API has actually completely changed since this PR was last published. I'm exporting the latest version now and am keen to get your take.

Reply:
Thank you for sharing the new version! It looks much cleaner, and you had found some interesting solutions.